crux_core/src/capability/executor.rs

```diff
@@ -151,37 +124,42 @@ impl QueuingExecutor {
         let context = &mut Context::from_waker(&waker);
 
         // poll the task
-        if task.as_mut().poll(context).is_pending() {
-            // If it's still pending, put the future back in the slot
-            self.tasks
-                .lock()
-                .expect("Task slab poisoned")
-                .get_mut(*task_id as usize)
-                .expect("Task slot is missing")
-                .replace(task);
-            RunTask::Suspended
-        } else {
-            // otherwise the future is completed and we can free the slot
-            self.tasks.lock().unwrap().remove(*task_id as usize);
-            RunTask::Completed
+        match task.as_mut().poll(context) {
+            Poll::Ready(command) => {
+                // otherwise the future is completed and we can free the slot
+                self.tasks.lock().unwrap().remove(*task_id as usize);
+                RunTask::Completed(command)
+            }
+            Poll::Pending => {
+                // If it's still pending, put the future back in the slot
+                self.tasks
+                    .lock()
+                    .expect("Task slab poisoned")
+                    .get_mut(*task_id as usize)
+                    .expect("Task slot is missing")
+                    .replace(task);
+                RunTask::Suspended
+            }
         }
     }
 }
 
-enum RunTask {
+enum RunTask<Event> {
     Missing,
     Unavailable,
     Suspended,
-    Completed,
+    Completed(Command<Event>),
```

Comment:
When a task resolves it always returns a command. 

Think of the existing model - e.g. an HTTP request that resolves to an event. When the request resolves, the resulting Event is implicitly spawned onto the executor through the `spawn_event` channel that is part of the capability context.

With the changes in this PR, the `spawn_event` channel has gone, and the event (or any other kind of `Command`) is returned directly from the resolved future. The command needs to be explicitly handled - either fed into `app.update` if it is an event, or spawned onto the executor if it is an effect (or else the command may be `Command::none()` in which case there is nothing left to do).

This happens in code via an interplay between `Core::process` and `QueueingExecutor::run_all`.

Reply:
Wait, are you saying the task completes every time the code in the future(s) hits an `.await` on one of the Crux specific interactions?

Reply:
So then I'm back to the question of how does it complete with a `Command` which is not just an `Event`? 

I'm not sure what that would mean in principle. The parent future of the effect task has finished, there are no more effects to do, the only possible outcomes are either nothing (we're done) or an event back into the app, right? What am I missing?

Reply:
It could be made to return `Option<Event>` instead, but you would some expressivity. 

A `Command` allows one to return any combination of events and effects to run next. This is useful e.g. to do an http request and a render call concurrently. You could do them in a single future with a `future::join` call, but then they are coupled together. Returning both with `Command::effect(render_fut).join(http_fut)` decouples them.

example: https://github.com/redbadger/crux/blob/03a7c6430745842eb6507c012824282b53e1f984/examples/cat_facts/shared/src/app.rs#L129-L131

Reply:
Actually possibly talking cross-purposes here. The above applies to the return type for `fn update(..)`. You want to know why any spawned future must return `Command`? The answer is it seems reasonable that you might want to spawn a future (or several futures) from within another future (emember we have no way to spawn futures than to return them as `context.spawn` has gone).

You might argue that futures shouldn't attempt to spawn futures directly, they should return events which kicks off the futures. Maybe true?

Reply:
I seee - if you wanted to start a separate task that runs concurrently without awaiting the result. Yea, that is something we need o retain as an ability...

Reply:
I suggest staring at https://github.com/redbadger/crux/blob/03a7c6430745842eb6507c012824282b53e1f984/examples/counter/shared/src/capabilities/sse.rs#L34-L65 until it makes sense :)

Reply:
> It's a stream where each output is itself a stream of Commands

That seems a step too far 😅 

I think the problem there starts with insisting on `Event` being a return, or generally the command async code working towards a final return value. I don't know that's the most straightforward approach to it.

The async code in capabilities (and now in `Command`) is quite intentionally pretty imperative, because it's "process" code, rather than a computation (as in - more Turing, less Church). It's much more naturally a "do this, then do that, then do that, then finish". And yes we _can_ model that with a recursion, but as can be seen in the code above^ it's really not very straightforward that way. We could do better with some monadic style combinators like `Command::event(event).and_then(...)` but... I think async is just a nicer interface for this type of work.